attention_and_transformers.ipynb

```diff
@@ -0,0 +1,4261 @@
+        "      })\n",
+        "      plotlosses.send()\n",
+        "      losses = []"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "IP7crkAoaqZ-"
+      },
+      "outputs": [],
+      "source": [
+        "data = next(train_dataset)\n",
+        "\n",
+        "predicted_string, encoder_weights, decoder_weights = get_model_prediction(params, data, True)\n",
+        "target_string =[ord_to_char(t) for t in data['target']]\n",
+        "input_string =[ord_to_char(t) for t in data['input']]\n",
+        "\n",
+        "for y_hat, y, input_ in zip(predicted_string[:2], target_string[:2], input_string[:2]):\n",
+        "      print(\"-\" * 10, 'Input', \"-\" * 11)\n",
+        "      print(input_)\n",
+        "      print(\"-\" * 10, 'Target', \"-\" * 10)\n",
+        "      print(y)\n",
+        "      print(\"-\" * 10, 'Prediction', \"-\" * 10)\n",
+        "      print(y_hat)\n",
+        "      print()\n",
+        "      print('*'*50)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "Fzq6UbqaACQv"
+      },
+      "outputs": [],
+      "source": [
+        "mean_encoder_weights = jnp.mean(encoder_weights,axis=0)\n",
+        "plot_weights(mean_encoder_weights, num_layers, num_heads, block='Encoder')"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "x_YmMCTLYYDU"
+      },
+      "outputs": [],
+      "source": [
+        "mean_decoder_weights = jnp.mean(decoder_weights,axis=0)\n",
+        "plot_weights(jnp.where(mean_decoder_weights==1, 0.5, mean_decoder_weights), num_layers, num_heads, block='Decoder')"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "35IPv-Ujjta4"
+      },
+      "source": [
+        "### Decoder / Encoder only models"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "yA6Kto9lCyRT"
+      },
+      "source": [
+        "In this practical, we introduced what is known as a vanilla (original) transformer, where a decoder and encoder work together. However, there is no need to have this architecture, as both the encoder and decoder blocks can also operate separately on their own. Most of today's well-known transformers fall into decoder/encoder-only transformers. \n",
+        "\n",
+        "We are usually interested in training our model to extract semantics or representations from our data when we use only encoder blocks. We are interested in generative models when we use decoder-only blocks (which in practice, is an encoder block with masks, i.e. not two MHA blocks). Famous models for all three cases are listed below. \n",
+        "\n",
+        "Vanilla transformers:\n",
+        "\n",
+        "* Original attention is all you need MLT\n",
+        "* T5\n",
+        "\n",
+        "Encoder only:\n",
```

Comment:
Aren't encoder-only and decoder-only models effectively the same (as there is no cross-attention in the decoder-only model)?

Reply:
@sebastianruder, I do mention in the notebook that they are essentially the same thing, but in many cases, the decoder-only models are trained with masks to block out future tokens to mimic the generative setting. 

I saw this distinction and terms in almost every single resource, so I am sticking with the convention here. 

But I do agree with you they are the same thing, and its the processing around them and how to use them that makes the difference :) 

Do you think this should rather be taken out @sebastianruder?